platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,233 @@
+import com.swirlds.platform.state.RosterStateAccessor;
+import com.swirlds.platform.state.RosterStateModifier;
+import com.swirlds.platform.state.signed.ReservedSignedState;
+import com.swirlds.platform.system.SoftwareVersion;
+import com.swirlds.state.spi.CommittableWritableStates;
+import com.swirlds.state.spi.WritableKVState;
+import com.swirlds.state.spi.WritableSingletonState;
+import com.swirlds.state.spi.WritableStates;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Read-write interface for accessing rosters states.
+ */
+public class WritableRosterStore implements RosterStateModifier {
+
+    private final WritableStates writableStates;
+    private final RosterStateAccessor rosterStateAccessor;
+    /**
+     * The roster state singleton. This is the state that holds the candidate roster hash and the list of pairs of
+     * active roster hashes and the round number in which those rosters became active.
+     *
+     * @implNote the use of {@link ReadablePlatformStateStore} and {@link WritablePlatformStateStore} to provide access
+     * to the roster states (beyond just the {@link PlatformState}) is deliberate, for convenience.
+     */
+    private final WritableSingletonState<RosterState> rosterState;
+
+    private final WritableKVState<ProtoBytes, Roster> rosterMap;
+
+    /**
+     * Constructs a new {@link WritableRosterStore} instance.
+     *
+     * @param writableStates the readable states
+     */
+    public WritableRosterStore(@NonNull final WritableStates writableStates) {
+        this.writableStates = writableStates;
+        this.rosterStateAccessor = new ReadableRosterStore(writableStates);
+        this.rosterState = writableStates.getSingleton(RosterStateId.ROSTER_STATES_KEY);
+        this.rosterMap = writableStates.get(RosterStateId.ROSTER_KEY);
+    }
+
+    /**
+     * Set the candidate roster if valid and doesn't yet exist in the state.
+     *
+     * @param candidateRoster a candidate roster to set.
+     */
+    @Override
+    public void setCandidateRoster(@NonNull final Roster candidateRoster) {
+        Objects.requireNonNull(candidateRoster);
+        RosterValidator.validate(candidateRoster);
+
+        // update the roster state
+        final RosterState currentRosterState = rosterStateOrThrow();
+        final Bytes candidateRosterHash = RosterUtils.hash(candidateRoster).getBytes();
+        final Builder rosterStateBuilder = RosterState.newBuilder()
+                .candidateRosterHash(candidateRosterHash)
+                .roundRosterPairs(currentRosterState.roundRosterPairs());
+        this.rosterState.put(rosterStateBuilder.build());
+
+        final Bytes currentCandidateRosterHash = currentRosterState.candidateRosterHash();
+        // remove existing candidate roster if present
+        this.rosterMap.remove(
```

Comment:
You should check whether the roster is in the roster pair list before removing from the map

Since this is also being done elsewhere in this class, consider a utility method to remove a given roster from the map, on the condition that it doesn't exist in the roster pair list

Reply:
How so? The roster pair list in the `RosterState` is a list of active rosters. This removal here is ensuring the not-yet-adopted previous candidate roster is being removed from the `RosterState` and `RosterMap`. The only way a CR makes its way into the roster pair list is when it gets adopted. At that time, its roster hash is removed from the `RosterState` and `candidateRosterHash` set to empty bytes. I don't see a need to check the roster pair list.